### PR TITLE
Develop validation rules

### PR DESCRIPTION
**Pull request by mialy-defelice**
This pull request allows for much more detailed specification of validation rules in the schema.

Design should enable developers to add additional rules in a modular way. Future rules to add could be a year validator and string length cap.

Raised errors should be informative enough to allow curator to easily locate and fix formatting.

_Taken from schema hub_ this PR will allow for the following validation rules:

**list:**
- **Single:**
   - Use list on its own to confirm that a comma separated list was provided by the user, and to convert the imported value (imported as a string) to an actual list. Values can come from Valid Values.
- **Multi:**
   - Can use list in conjunction with regex rule to validate that the items in a list follow a certain pattern. How it works now is all the values in the list would need to follow the same pattern. This is ideal for when users need to provide a list of IDs.
     - Format:
         - `list :: regex module regular_expression`
         - ‘::’ delimiter used to separate list from the regex rule
     - Example:
         - `list :: regex search [HTAN][0-9]{1}_[0-9]{4}_[0-9]*`

**regex:** 
- Use regex validation rule when you want to require that a user input values in a specific format, i.e. an ID that follows a particular format. 
- Note: in a future PR #510 if module is specified as `match` regex validation (along with strictness) can be applied to google sheets. Use `match` with care and only if your regex calls for it.
  - Format: 
      - `regex module regular_expression`
      - Module: is the re module that you want to use. A common one would be search. Refer to python re source material to find the most appropriate module to use.
      - Single spaces separate the three strings.
  - Example:
     -` regex search [0-9]{4}\/[0-9]*`
     - The regular expression defined above allows comparison to an expected format of a histological morphology code.

**type:**
- Use the type checker by setting the ‘Validation Rules’  to one of the types listed below. No additional specifiers are allowed. 
   - **float:**
       - Use in the case that the value entered needs to be a float. Checks that the value entered for is a float (ie. a number that is not an interger).
   - **int:**
      - Use in the case that the value entered needs to be an interger. Checks that the value entered for is an interger.
   - **num:**
      - Use in the case that the value entered just needs to be a numeral and does not need to be specified as an interger or float. Checks that the value entered is a numeral.
   - **str:**
      - Use in the case that the value entered needs to be a string. Checks that the value entered is a string.
      - Future feature is a string of a specified length.

**url:**
   - Setting ‘Validation Rules” to URL implies the user should add a URL to a free text box as a string. This function will check that the user has provided an actual, usable URL. It will check for any standard URL error and throw an error if one is found. Further additions to this rule can allow for checking that a specific type of URL is added. For example, if the user needs to add a protocols.io URL, ‘protocols.io’ can be added after url to perform this check. If the provided url does not contain this specific string, an error will be raised.
   - Format:
       - url must be specified first then an arbitrary number of strings can be added after (separated by spaces) to add additional levels of specificity.
       - Alternatively, its valid to pass only ‘url’ to simply check if the input is a url.
       - Examples:
          - `url protocols.io`
            - Will check that any input is a valid URL, and will also check to see that the URL contains the string ‘protocols.io’. If not, an error will be raised.
          - ` url dx.doi protocols.io`
            - Will check that any input is a valid URL, and will also check to see that the URL contains the strings  ‘dx.doi’ and ‘protocols.io’. If not, an error will be raised.

### CODE REVIEW

**Comment by mialy-defelice** [comment]
```
# generating sheet api request to populate a dropdown or a multi selection UI 
if len(req_vals) > 0 and not "list" in validation_rules:

# if more than 0 values in dropdown use ONE_OF_RANGE type of validation since excel and openoffice
# do not support other kinds of data validation for larger number of items (even if individual items are not that many
# excel has a total number of characters limit per dropdown...)

validation_body = self._get_column_data_validation_values(
spreadsheet_id, req_vals, i, validation_type="ONE_OF_RANGE"
)
elif "list" in validation_rules:
# if list is in validation rule attempt to create a multi-value
# selection UI, which requires explicit valid values range in
# the spreadsheet
validation_body = self._get_column_data_validation_values(
spreadsheet_id,
req_vals,
i,
custom_ui=False,
input_message="",
validation_type="ONE_OF_RANGE",
)
```

Just realized this will be an issue with list validator. I think this other version of using list should be moved out of it being a validation_rule. Since it's not so much a rule, as an action that needs to be performed to create the sheet properly.

**Comment by mialy-defelice** [comment]
Possible simple solution to above problem call old type of `list` `vv_list`

**Review by BrunoGrandePhD — COMMENTED** [review]
I'm excited for this feature! I don't have time to test it, but I took a quick look at the code. I just had a few suggestions. Otherwise, this looks great! 🚀

File: schematic/models/validate_attribute.py [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+        Purpose:
+            Determine if values for a particular attribute are comma separated.
+        Input:
+            - val_rule: str, Validation rule
+            - manifest_col: pd.core.series.Series, column for a given attribute
+        Returns:
+            - manifest_col: Input values in manifest arere-formatted to a list
+            - Error log, error list
+        '''
+        def generate_list_error(list_string: str, row_num: str,
```
[the remark this answers is not in the document]
Comment (mialy-defelice) [reply schematic/models/validate_attribute.py:41]: Moved on their own.

File: schematic/models/metadata.py [hunk 2]
```diff
@@ -250,37 +256,14 @@ def validateModelManifest(
...
-
-            for error in sorted(v.iter_errors(annotation), key=exceptions.relevance):
-                errorRow = i + 2
-                errorCol = error.path[-1] if len(error.path) > 0 else "Wrong schema"
-                errorMsg = error.message[0:500]
-                errorVal = error.instance if len(error.path) > 0 else "Wrong schema"
-
-                errors.append([errorRow, errorCol, errorMsg, errorVal])
+        # Validate Manifest Rules
+        manifest, vmr_errors = ValidateManifest.validate_manifest_rules(self, manifest, self.sg)
```
[the remark this answers is not in the document]
Comment (mialy-defelice) [reply schematic/models/metadata.py:260]: Great suggestion! Created `validate_all`

**Comment by mialy-defelice** [comment]
All comments have been addressed :) Ready to be merged.

**Comment by milen-sage** [comment]
@mialy-defelice just tested this PR and it works nicely!

**Comment by milen-sage** [comment]
@mialy-defelice could you take a look at the merge conflicts in schematic/manifest/generator.py^? Looks like a straightforward resolution, but you'd be best to confirm. I think we're ready to merge after that! 

Meanwhile, I'm going through #510.

File: config.yml [hunk 3]
```diff
@@ -1,26 +1,30 @@
 # Do not change the 'definitions' section unless you know what you're doing
 definitions:
-  synapse_config: ".synapseConfig"
-  creds_path: "credentials.json"
-  token_pickle: "token.pickle"
-  service_acct_creds: "schematic_service_account_creds.json"
+  synapse_config: "/Users/mialydefelice/Documents/schematic/schematic_package_docs/.synapseConfig"
```
Comment (milen-sage) [inline config.yml:3]: @mialy-defelice could you recommit the example config file?

**Review by milen-sage — CHANGES_REQUESTED** [review]
@mialy-defelice looks good! Just one comment around the config file - it's your personal one, which includes paths on your machine.

**Review by milen-sage — APPROVED** [review]

**Pull request merged**